internal/rbd/controllerserver.go

```diff
@@ -743,9 +743,51 @@ func checkContentSource(
 	return nil, nil, status.Errorf(codes.InvalidArgument, "not a proper volume source")
 }
 
+// checkErrAndUndoReserve work on error from genVolFromVolID() and undo omap reserve.
+func (cs *ControllerServer) checkErrAndUndoReserve(
+	ctx context.Context,
+	err error,
+	volumeID string,
+	rbdVol *rbdVolume, cr *util.Credentials) (*csi.DeleteVolumeResponse, error) {
+	if errors.Is(err, util.ErrPoolNotFound) {
+		log.WarningLog(ctx, "failed to get backend volume for %s: %v", volumeID, err)
+
+		return &csi.DeleteVolumeResponse{}, nil
+	}
+
+	// if error is ErrKeyNotFound, then a previous attempt at deletion was complete
+	// or partially complete (image and imageOMap are garbage collected already), hence return
+	// success as deletion is complete
+	if errors.Is(err, util.ErrKeyNotFound) {
+		log.WarningLog(ctx, "Failed to volume options for %s: %v", volumeID, err)
+
+		return &csi.DeleteVolumeResponse{}, nil
+	}
+
+	// All errors other than ErrImageNotFound should return an error back to the caller
+	if !errors.Is(err, ErrImageNotFound) {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
```

Comment:
there are three initial error checks, might be cleaner to use a switch-statement?

Reply:
imo, the current if check is sufficient.